### DISCUSSION TITLE
berry -s -g solidify_all.be error: internal_error: Strings greater than 255 chars not supported yet

### DISCUSSION THREAD

**Discussion by lboue** [arendst/Tasmota#20469] [Q&A] [answered]
Hello,

I am trying to solidify the Matter berry part of the firmware to add a new Matter device type and I get this error:
`internal_error: Strings greater than 255 chars not supported yet`

Can you help me find a workaround?
Dev branch: https://github.com/lboue/Tasmota-1/tree/device/water_leak

```
gitpod /workspace/Tasmota-1/lib/libesp32/berry_matter (device/water_leak) $ ../berry/berry -s -g solidify_all.be
Parsing:  Matter_0_Inspect.be
Parsing:  Matter_Base38.be
Parsing:  Matter_Commissioning.be
...
Parsing:  Matter_Plugin_9_Virt_Sensor_Occupancy.be
Parsing:  Matter_Plugin_9_Virt_Sensor_Pressure.be
Parsing:  Matter_Plugin_9_Virt_Sensor_Temp.be
Parsing:  Matter_Plugin_9_Virt_Sensor_Waterleak.be
Parsing:  Matter_Plugin_z_All.be
Parsing:  Matter_Profiler.be
Parsing:  Matter_Session.be
Parsing:  Matter_Session_Store.be
Parsing:  Matter_TCP_async.be
Parsing:  Matter_TLV.be
Parsing:  Matter_UDPServer.be
Parsing:  Matter_UI.be
internal_error: Strings greater than 255 chars not supported yet
stack traceback:
        <native>: in native function
        <unknown source>: in function `parse_file`
        <unknown source>: in function `main`
```

[solidify_all.be.log](https://github.com/arendst/Tasmota/files/13906737/message.txt)

**Comment by sfromis** [comment]
Your addition resulted in a too long string, with `_CLASSES_TYPES` growing from 251 to 273 chars, just sneaking over the mentioned restriction. Not tried this in a solidification context, but I think that you could get around the limitation by making it into two strings by using a `+` concatenation operation once, instead of only relying on the implicit concatenation performed by the Berry compiler when seeing two strings without any operator.

**Reply by benjaminnestler** [reply]
I've tried it as following:

```berry
static var _CLASSES_TYPES_LOCAL = "|relay|light0|light1|light2|light3|shutter|shutter+tilt"
                              "|temperature|pressure|illuminance|humidity|occupancy|onoff|contact|flow|thermostat"

  static var _CLASSES_TYPES_VIRTUAL = "|-virtual|v_relay|v_light0|v_light1|v_light2|v_light3"
                              "|v_temp|v_pressure|v_illuminance|v_humidity|v_occupancy|v_contact|v_flow|v_thermostat"
  static var _CLASSES_TYPES = self._CLASSES_TYPES_LOCAL + self._CLASSES_TYPES_VIRTUAL
```
That leads to:
```
syntax_error: string:40: 'self' undeclared (first use in this function)
stack traceback:
        <native>: in native function
        <unknown source>: in function `parse_file`
        <unknown source>: in function `main`
```

So that didn't worked. I think due to the `static var definition?

**Reply by sfromis** [reply]
That was not what I meant, which instead was to just insert one single `+` at a spot in the multiple lines of `_CLASSES_TYPES`.

Your less straightforward approach would require you to replace `self` by the class name, indeed due to it being static vars

In both cases, I've not tried if an expression even works for static var initialization.

**Reply by benjaminnestler** [reply]
> Your less straightforward approach would require you to replace self by the class name, indeed due to it being static vars

Ahh okay. Thanks for your hint.

> In both cases, I've not tried if an expression even works for static var initialization.

Both cases didn't worked. The error is the same.

**Comment by s-hadinger** [comment]
I will try to add solidification of long strings in Berrt

**Reply by benjaminnestler** [reply]
Is there real restriction using longer strings during solidification?

The internal error will be raised by this:

```c
const char * str = str(var_tostr(value));
            size_t len = strlen(str);
            if (len >= 255) {
                be_raise(vm, "internal_error", "Strings greater than 255 chars not supported yet");
            }
            size_t id_len = toidentifier_length(str);
            char id_buf[id_len];
            toidentifier(id_buf, str);
            if (!str_literal) {
                logfmt("be_nested_str(%s)", id_buf);
            } else {
                logfmt("be_nested_str_weak(%s)", id_buf);
            }
```

The `size_t` type and the `toidentifier_length()` function seems to be well-prepared for longer strings.

The `logfmt()` macro uses a limited buffer of `INST_BUF_SIZE 768`. Is this the magical limit for the whole solidification of strings.

**Reply by sfromis** [reply]
While the coding in `toidentifier` may not have a hard limit, my assumption would be that the identifier as such cannot be more than 255 chars, which is a common max length for single-part names.

Found this bit about c++ identifiers:
> An identifier is a sequence of characters, of up to a certain length (compiler-dependent, typically 255 characters), comprising uppercase and lowercase letters (a-z, A-Z), digits (0-9), and underscore "_".

Question may be if the compiler in use is less restrictive, or even if it relies on being "allowed" to only see the first 31/63 chars as significant.

**Reply by lboue** [reply]
Thanks

**Reply by benjaminnestler** [reply]
> Question may be if the compiler in use is less restrictive, or even if it relies on being "allowed" to only see the first 31/63 chars as significant.

Seems to be that only a Standard C (not C++)  issue. I didn't found a significant length limit for identifier names in this in cpp reference book:
https://en.cppreference.com/w/cpp/language/identifiers

The C Standard limits the identifier names (see https://en.cppreference.com/w/c/language/identifier)

> Even though there is no specific limit on the length of identifiers, early compilers had limits on the number of significant initial characters in identifiers and the linkers imposed stricter limits on the names with [external linkage](https://en.cppreference.com/w/c/language/storage_duration). C requires that at least the following limits are supported by any standard-compliant implementation:

<table>
    <thead>
        <tr>
            <th>C Standard</th>
            <th>Restrictions</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td rowspan=1 align="center">until C99</td>
            <td rowspan=1 align="left"><ul><li>31 significant initial characters in an internal identifier or a macro name</li>
<li>6 significant initial characters in an external identifier</li>
<li>511 external identifiers in one translation unit</li>
<li>127 identifiers with block scope declared in one block</li>
<li>1024 macro identifiers simultaneously defined in one preprocessing translation unit</li></u></td>
        </tr>
        <tr>
            <td rowspan=1 align="left">since C99</td>
            <td align="left"><ul><li>63 significant initial characters in an internal identifier or a macro name</li>
<li>31 significant initial characters in an external identifier</li>
<li>4095 external identifiers in one translation unit</li>
<li>511 identifiers with block scope declared in one block</li>
<li>4095 macro identifiers simultaneously defined in one preprocessing translation unit</li></u></td>
        </tr>
    </tbody>
</table>

**Reply by s-hadinger** [reply]
I'm working on it and progressing. The good news is that gcc has no limit in identifier size so I will follow the same scheme for simplicity. Solidification of long string are already a rare stuff, not being able to solidify very long strings with non gcc compilers will be a documented limitation

**Reply by sfromis** [reply]
> Seems to be that only a Standard C (not C++) issue.

The page where I found that text is specifically about C++, but 10 years ago. And it is plausible that the solidification limit of 255 was set to have something not at risk of causing trouble with any compiler, or maybe due to old habits.

**Reply by benjaminnestler** [reply]
> And it is plausible that the solidification limit of 255 was set to have something not at risk of causing trouble with any compiler, or maybe due to old habits.

Sure you're right. Better to be on the safe side.

**Reply by s-hadinger** [reply]
Actually the limit of 255 is unrelated. Berry stores all strings less than 255 bytes in a giant hashmap to avoid duplication of the same strings. But above 256 bytes, it does not care anymore and makes the assumption that creating twice the same string above 255 bytes in of low probability

**Reply by s-hadinger** [reply]
This is why I didn't include large strings in the solidification process because I thought (wrongly) that we didn't care. But there is no fundamental blocker

**Reply by benjaminnestler** [reply]
> I'm working on it and progressing. The good news is that gcc has no limit in identifier size so I will follow the same scheme for simplicity. Solidification of long string are already a rare stuff, not being able to solidify very long strings with non gcc compilers will be a documented limitation

Regarding the specific string `_CLASS_TYPES` (and `_CLASS_TYPES2`) @lboue speaks of, what do you think of using a list, with the name of the available classes as list items and change the functions `show_plugins_hints_js()` and `plugin_option()` which will currently extracting the class names into a list.

Like this:
```berry
static var _CLASSES_TYPES_LIST = [
"relay","light0","light1","light2","light3","shutter","shutter_and_tilt"
"temperature","pressure","illuminance","humidity","occupancy","onoff","contact","flow",
"thermostat_tuya",
"-virtual","v_relay","v_light0","v_light1","v_light2","v_light3",
"v_temp","v_pressure","v_illuminance","v_humidity","v_occupancy","v_contact","v_flow"
]
```

```berry
def show_plugins_hints_js(class_list)
    import webserver
    import json
    import string

    #var class_types = class_list
    #for cl: class_list
    #  class_types += string.split(cl, '|')
    #end

    var hm = {}
    var hl = []

    for typ: class_list
          # original code
```

```berry
def plugin_option(cur, class_list)
    import webserver
    import string
    
    #var class_types = []
    #for cl: class_list
    #  class_types += string.split(cl, '|')
    #end

    var i = 0
    while i < size(class_list)
      var typ = class_list[I]
      # original code
```

**Reply by Jason2866** [reply]
We don't use only gcc for solidification. clang is used in GH Actions, since gcc produced sometimes! strange errors.

**Reply by s-hadinger** [reply]
@benjaminnestler Using a list takes significantly more flash space.

I have a working version of solidification of long strings, I hope to be able to push it today.

**Reply by s-hadinger** [reply]
I tried with an identifier of 7600 bytes, and the compiler/linker didn't complain

**Comment by s-hadinger** [comment]
@lboue There is limit anymore

**Reply by sfromis** [reply]
`s/limit/no limit/`

**Reply by lboue** [reply, marked as answer]
@s-hadinger 
Thanks for your help and commits. I'll be able to resume my tests 🙂
